File: 1077.py
def main():
    n = int(input())

    while (n):
        infix = input()
        solve(infix)
        n = n - 1

def solve(infix):
    ops = {'+': 1, '-': 1, '*': 2, '/': 2, '^': 3}
    t = []
    for c in infix:
        if (c == '('):
            t.append(c)
        elif (c == ')'):
            ti = t.pop()
            while (t and ti != '('):
                print(ti, end='')
                ti = t.pop()
        elif (c in ops):
            if (not t):
                t.append(c)
            else:
                ti = t[len(t)-1]
                if (ti == '('):
                    t.append(c)
                elif (ops[c] > ops[ti]):
                    t.append(c)
                else:
                    while (t and ops[ti] >= ops[c]):
                        print(ti, end='')
                        t.pop()
                        if (t):
                            ti = t[len(t) - 1]
                            if (ti == '('):
                                break
                    t.append(c)
        else:
            print(c, end='')
    while (t):
        print(t.pop(), end='')
    print()


if __name__ == '__main__':
    main()

File: 1077.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(0);

  map<char, int> ops{{'-', 1}, {'+', 1}, {'*', 2}, {'/', 2}, {'^', 3}};
  stack<char> t;
  string s;

  int n;
  cin >> n;
  while (n--) {
    cin >> s;
    for (int i = 0; i < s.size(); i++) {
      char c = s[i];
      if (c == '(') {
        t.push(c);
        continue;
      }
      if (c == ')') {
        char ti = t.top();
        while (ti != '(') {
          cout << ti;
          t.pop();
          ti = t.top();
        }
        t.pop();
        continue;
      }
      auto search = ops.find(c);
      if (search != ops.end()) { // it is an operator
        if (t.empty()) {
          t.push(c);
        } else {
          char ti = t.top();
          if (ti == '(') {
            t.push(c);
          } else if (ops[c] > ops[ti]) {
            t.push(c);
          } else {
            while (ops[ti] >= ops[c]) {
              cout << ti;
              t.pop();
              if (t.empty()) {
                break;
              }
              ti = t.top();
              if (ti == '(') {
                break;
              }
            }
            t.push(c);
          }
        }
      } else {
        cout << c;
      }
    }
    while (!t.empty()) {
      cout << t.top();
      t.pop();
    }
    cout << endl;
  }
  return 0;
}
